src/lsp/cobol_parser/expect/gen_expect.ml

```diff
@@ -167,234 +175,238 @@ module CompEntrySet = Set.Make(struct
+    | T _, _, _ -> begin
         match continue_while_terminal rhs idx [] with
-          | [] -> []
-          | l -> [K (NEL.of_list l)] end
-      | exception Invalid_argument _ -> []
+        | [] -> []
+        | l -> [K (NEL.of_list l)] end
+    | exception Invalid_argument _ -> []
   in
   Lr1.tabulate begin fun lr1 ->
     let custom_comp_entries = CompEntrySet.from_list_map
-      ~add:CompEntrySet.add_option
-      begin function
-        | N nonterm, _ -> nonterminal_filter_map nonterm
-        | _ -> None end
-      (Lr1.transitions lr1) in
+        ~add:CompEntrySet.add_option
+        begin function
+          | N nonterm, _ -> nonterminal_filter_map nonterm
+          | _ -> None end
+        (Lr1.transitions lr1) in
     CompEntrySet.from_list_map
       ~init:custom_comp_entries ~add:CompEntrySet.add_list
       next_tokens_from_item (Lr0.items @@ Lr1.lr0 lr1)
   end
 
 let emit_acceptable_terminals_in_env ppf = (* taking transitions *)
-  Fmt.pf ppf {|
+  Fmt.string ppf {|
 let acceptable_terminals_in ~env : completion_entry list =
   NEL.(match Menhir.current_state_number env with
 |};
   Lr1.fold (fun lr1 acc ->
-    match completion_entries_of lr1 with
+      match completion_entries_of lr1 with
       | s when CompEntrySet.is_empty s -> acc
       | s -> CompEntryMap.add_to_list s lr1 acc)
-  CompEntryMap.empty
+    CompEntryMap.empty
   |> inv (CompEntryMap.fold (fun s l acc -> (l, CompEntrySet.elements s)::acc)) []
   |> pp_match_cases ppf
     Fmt.(using Lr1.to_int int)
     (pp_brackets @@ pp_list pp_completion_entry)
     "[])"
 
 let guess_default_value ~mangled_name ~typ =
-  match typ, mangled_name with
-  | _ when EzString.ends_with ~suffix:"option" typ ||
-          EzString.starts_with ~prefix:"option_" mangled_name
-  -> Some "None"
-  | _ when EzString.ends_with ~suffix:"list" typ ||
-  EzString.starts_with ~prefix:"loption_" mangled_name ||
-  EzString.starts_with ~prefix:"list_" mangled_name
-  -> Some "[]"
-  | _ when EzString.ends_with ~suffix:"unit" typ
-  -> Some "()"
-  | _ when EzString.ends_with ~suffix:"bool" typ ||
-  EzString.starts_with ~prefix:"boption_" mangled_name
-  -> Some "false"
-  | _ -> None
+  if EzString.ends_with ~suffix:" option" typ ||
+     EzString.starts_with ~prefix:"option_" mangled_name
+  then Some "None"
+  else if EzString.ends_with ~suffix:" list" typ ||
+          EzString.starts_with ~prefix:"loption_" mangled_name ||
+          EzString.starts_with ~prefix:"list_" mangled_name
+  then Some "[]"
+  else if String.equal typ "unit"
+  then Some "()"
+  else if String.equal typ "bool" ||
+          EzString.starts_with ~prefix:"boption_" mangled_name
+  then Some "false"
+  else None
 
 let is_nullable_without_recovery = Nonterminal.tabulate begin fun nt ->
-  Nonterminal.nullable nt &&
+    Nonterminal.nullable nt &&
     Nonterminal.attributes nt
     |> List.find_opt (Attribute.has_label "recovery")
```

Comment:
Do you really need to deal with `recovery` attributes here? Wouldn't this be given by `default_value` anyways?

Reply:
Well yes but no need to create an entry with a guessed type in the function if we know the case will be caught by the recovery, this would only add to the risk of wrongly guessing the type...

Reply:
I think that should be the other way round: the function that guesses the type should be a sort of under-approximation, i.e, return a result only if it is sure that is correct. The way to do that is probably by working only on the type string, and checking some basic regexps. For instance if the type string matches  `[^ ]+ option`, then it is surely a `None`. That's ok if less expressive cases require a `@default` attribute.

Apologies if I lead you the wrong way with checks for suffixes. Here you may just `String.split_on_char ' '` and check the last element is `"option"` or `"list"`. Or use regexps. That's up to you.

With a sound guess, we can leave the remaining cases to the `Defaults` module (effectively the recovery).